mesonbuild/modules/python.py

```diff
@@ -276,11 +276,47 @@ def set_env(name, value):
 
 
 INTROSPECT_COMMAND = '''\
+import os.path
 import sysconfig
 import json
 import sys
 
-install_paths = sysconfig.get_paths(vars={'base': '', 'platbase': '', 'installed_base': ''})
+def get_distutils_paths(*, scheme=None, prefix=None):
+    import distutils.dist
+    distribution = distutils.dist.Distribution()
+    install_cmd = distribution.get_command_obj('install')
+    if prefix is not None:
+        install_cmd.prefix = prefix
+    if scheme:
+        install_cmd.select_scheme(scheme)
+    install_cmd.finalize_options()
+    return {
+        'data': install_cmd.install_data,
+        'include': os.path.dirname(install_cmd.install_headers),
+        'platlib': install_cmd.install_platlib,
+        'purelib': install_cmd.install_purelib,
+        'scripts': install_cmd.install_scripts,
+    }
+
+# On Debian derivatives, the Python interpreter shipped by the distribution uses
+# a custom install scheme, deb_system, for the system install, and changes the
+# default scheme to a custom one pointing to /usr/local and replacing
+# site-packages with dist-packages.
+# See https://github.com/mesonbuild/meson/issues/8739.
+# XXX: We should be using sysconfig, but Debian only patches distutils.
+
+debian_distutils = False
+if sys.version_info < (3, 12):
```

Comment:
That's doing a gamble long into the future, we are 3.9 currently... anything can still happen, no? What about instead:
```python
try:
  import distutils.command.install
  debian_distutils = 'deb_system' in distutils.command.install.INSTALL_SCHEMES
except:
  pass
```
As long as `deb_system` remains in distutils, it's pretty safe to assume they kept the patch. That would keep working if they postpone migration to >3.12.

Reply:
`distutils.command.install` is not installed on Debian by default, it is provided by `python3-distutils`, so we want to actually import and error out, perhaps with a nicer message.

> As long as `deb_system` remains in distutils, it's pretty safe to assume they kept the patch. That would keep working if they postpone migration to >3.12.

They won't, that would require the Steering Council to revert the PEP (https://www.python.org/dev/peps/pep-0632/), which is very very unlikely. `distutils` has been frozen in CPython and extracted to https://github.com/pypa/distutils/. It will then be merged into setuptools at some point in the future.

Reply:
> `distutils.command.install` is not installed on Debian by default, it is provided by `python3-distutils`, so we want to actually import and error out, perhaps with a nicer message.

Yes, that's the whole problem. Meson must not depend on extra modules, so we cannot fail here, we have to fallback with a warning (hinting at install distutils). Pretty sure there are tons of CI out there that only install `python3` package in their docker image and we cannot break them.

Reply:
This script already depends on distutils for `links_against_libpython`.

> Pretty sure there are tons of CI out there that only install `python3` package in their docker image and we cannot break them.

I'd guess almost everyone has `python3-pip`, which pull `python3-distutils`.

Reply:
tbh, this whole story is pretty much going full circle back to just hardcode the path on debian systems. If python3-distutils is not installed by default we still have to hardcode the path... The tricky part is to correctly detect when debian will stop needing that hack, and I have sub-zero confidence with that being 3.12, that's the kind of break everything plan that gets postponed for years.

Reply:
> This script already depends on distutils for `links_against_libpython`.

Hmm, that's actually a good point... probably make that conditional too. After double checking, that will make the script fail, but Meson then continue with some default values. IMHO we should not make that script abort, but return sysconfig values instead.

Reply:
> I'd guess almost everyone has `python3-pip`, which pull `python3-distutils`.

On developer's laptop of course, but here I'm more thinking about build bots, ci docker, etc...

Reply:
distutils is not an external module, it's part of the standard library, Debian just splits it into a different package against the upstream wishes.

The Meson package in Debian should depend on `python3-disutils`, and if people installing Meson via pip, well, they have pip so they should have distutils.

IMO it's reasonable to ask people to install `python3-distutils` if it is not there for some reason, given it should already be there anyway. This should not affect almost anyone.

Reply:
Maybe I'm being too cautious... `import('python').find_installation() aborts already in a debian/ubuntu docker where I install only meson package:
```
meson.build:3:17: ERROR: <ExternalProgram 'python3' -> ['/usr/bin/python3']> is not a valid python or it is missing setuptools
```

Been like that for 3 years, no complains... ok, I guess we could do it then...

Reply:
I think you should report a bug to the distro, the Meson package should depend on `python3-distutils`.

Reply:
Actually, people do complain, but we just put the blame on Debian for their mess, it seems: https://github.com/mesonbuild/meson/issues/4267.

Reply:
Anyway, just let me know if you want to make any change to the logic.

Reply:
ok, ok, sorry for the long conversation, I learned a lot today :D

Still, I'm not sure we should make this `<3.12` check. What happens if 3.12 gets released and debian still haven't patched sysconfig instead of distutils? Why isn't it enough to check if `deb_system` is there regardless of the version?

Reply:
No worries, this is a mess :sweat_smile: 

Considering the script will fail on `links_against_libpython` if distutils is missing anyway, we can do that. The worry was that the import would fail if distutils is missing and we would mistakenly think that we are in a new version of Python without it. But yeah, we are gonna fail anyway later.

Reply:
No worries, this is a mess :sweat_smile: 

Considering the script will fail on `links_against_libpython` if distutils is missing anyway, we can do that. The worry was that the import would fail if distutils is missing and we would mistakenly think that we are in a new version of Python without it. But yeah, we are gonna fail anyway later.

Reply:
I added a heuristic to keep the helpful message.

Reply:
But one worry about this is that after 3.12, people will be able to provide a `distutils` module via pip. So yeah, I would prefer the version check, but we can go with the try block if you feel it is better.

Reply:
Yeah, indeed meson has always hard errored if:
- the python installation is required
- distutils is not available

and with an error message that mentioned setuptools :( but that got fixed in #9196

Since INTROSPECT_COMMAND already fails with a traceback when distutils is installed, and this traceback is used to `raise mesonlib.MesonException(f'{python} is not a valid python or it is missing distutils')`, I don't see any point in adding a bunch of complicated try/excepts here. Please just import `distutils.command.install` on one line and check for 'deb_system' in INSTALL_SCHEMES on another line, and defer to the existing handling to raise a suitable error message.

Reply:
Well, this identifies if the origin of the issue is indeed python3-distutils missing on Debian, and points that out in the traceback, so I think there is a point in keeping it. But I can remove if you want.

Reply:
Maybe the existing error message can be improved to provide a debian-specific recommendation?

Note that the traceback is run in a subprocess and then logged as mlog.debug in meson itself, and so is your ModuleNotFoundError. However the MesonException is logged as a fatal error. :)

Reply:
Can we do that in a separate PR? I really want to move this forward as it is blocking my work. With the patch as-is, user will get the same message they were getting previously, which the added extra that they now get a better debug message.

Reply:
It should not be that complicated:

```diff
diff --git a/mesonbuild/modules/python.py b/mesonbuild/modules/python.py
index bae5f9f05..f34d56d94 100644
--- a/mesonbuild/modules/python.py
+++ b/mesonbuild/modules/python.py
@@ -281,6 +281,8 @@ import sysconfig
 import json
 import sys
 
+import distutils.command.install
+
 def get_distutils_paths(*, scheme=None, prefix=None):
     import distutils.dist
     distribution = distutils.dist.Distribution()
@@ -305,22 +307,7 @@ def get_distutils_paths(*, scheme=None, prefix=None):
 # See https://github.com/mesonbuild/meson/issues/8739.
 # XXX: We should be using sysconfig, but Debian only patches distutils.
 
-try:
-    # Debian patches distutils, keeping only distutils.version, so
-    # `import distutils` will always work, but
-    # `import distutils.command.install` will fail if python3-distutils is
-    # not installed. Let's try detecting that and error out with an helpful
-    # error message.
-    import distutils
-    try:
-        import distutils.command.install
-    except ModuleNotFoundError:
-        raise ModuleNotFoundError('Unable to import distutils, please install python3-distutils')
-    debian_distutils = 'deb_system' in distutils.command.install.INSTALL_SCHEMES
-except ModuleNotFoundError:
-    debian_distutils = False
-
-if debian_distutils:
+if 'deb_system' in distutils.command.install.INSTALL_SCHEMES
     paths = get_distutils_paths(scheme='deb_system')
     install_paths = get_distutils_paths(scheme='deb_system', prefix='')
 else:
@@ -719,7 +706,11 @@ class PythonModule(ExtensionModule):
                 return python
             else:
                 if required:
-                    raise mesonlib.MesonException(f'{python} is not a valid python or it is missing distutils')
+                    msg = f'{python} is not a valid python or it is missing distutils.'
+                    if mesonlib.is_debianlike():
+                        msg += ' Try installing python3-distutils.'
+                    raise mesonlib.MesonException(msg)
+
                 return NonExistingExternalProgram()
 
         raise mesonlib.MesonBugException('Unreachable code was reached (PythonModule.find_installation).')
```


The final diff hunk is complete overkill, the lack of distutils in versions of python released before the end of 2023 is already a sign that is_debianlike() is true... but why not.

Reply:
> the lack of distutils in versions of python released before the end of 2023 is already a sign that is_debianlike() is true

But the opposite is not true. Your patch will fail with the `Try installing python3-distutils` message if the introspection script fails on Python installations that are not patched by Debian, as long we are in a Debian based system. This is the case of the ubuntu runner in Github Actions for example, where they have their own Python.

It's not a big deal in this case in particular, but it is precisely this kind of inaccuracies, well it was exactly this inaccuracy, that lead to bugs like the one that is being fixed in this PR. For that reason I always like to avoid such inaccurate checks, even if they do not have significant consequences -- they often end up leaking to places that do have consequences.

Anyway, your patch completely removes the accurate pinpoint of the issue. The approach is basically "the script failed? well, maybe install pyhon3-distutils idk".

I don't really want to keep arguing with you, so just fix up my commits with the code you deem to be better 🙃

Reply:
Both messages are "the script failed? well maybe install python3-distutils idk".

And the only other failure mode is a `python2` executable reporting it is >=3.0 (or vice versa).

The difference between the two messages is that one of them actually prints output which the user will see, and one of them logs debug output which will *only ever* be seen by people who already know what it means.

> well it was exactly this inaccuracy, that lead to bugs like the one that is being fixed in this PR

No? The inaccuracy which this PR is intending to fix is:

meson uses incorrect sources of information for one vendor's custom-patched runtime, instead let us detect and special case this to handle it correctly.

The inaccuracy I am proposing is:

there is currently a vague 'idk' message, and the only way to improve it is to improve *that* message, which you can do by implementing subprocess state communication or, if you "really want to move this forward as it is blocking my work", you can at least improve the message to mention the package name.

Reply:
> meson uses incorrect sources of information for one vendor's custom-patched runtime, instead let us detect and special case this to handle it correctly.

The incorrect source of information is checking for the presence of `/etc/debian-version` and deciding that that means the Python installation is patched by Debian.

Like I said, I don't want to argue, so just fix up my commit, or push a new commit with your patch.

Reply:
An alternative way to phrase my review comment on this import nest:

This is logged at mlog.debug which is the wrong place to log informative guidance. No one will ever see it, it only exists so we can tell people "post this really long logfile for us to interpret". We can just get rid of it entirely as it effectively serves no purpose other than to make the code complex.

Aside: feel free to improve this other line of code which is user-visible, now or in the future.

Reply:
Yes, and I agree. I asked above if we could do this in a different PR so that we can move this one, I expect to be some discussion on how exactly it is reported.

Reply:
Maybe we could just remove the try/except stuff (diff hunks 1 and 2 which I pasted above, but not diff hunk 3), not touch the error message or try to do any error reporting ***yet***, and open a draft PR with the `ModuleNotFoundError` added back and where the pros and cons of different forms of error reporting could be discussed.